### PR TITLE
EdStem API & macros

### PR DESCRIPTION
**Pull request by TrueshotBarrage**
Summary:
- (Magd) New EdStem API & macros
- (David) Added recursive cleanup macro for EdStem threads.

Examples:

We start with this Ed thread:
image

Then we make a request using our API...
```json
{'thread': {'id': 277600, 'user_id': 147342, 'course_id': 4757, 'editor_id': None, 'accepted_id': 659056, 'duplicate_id': None, 'number': 12, 'type': 'question', 'title': 'use of int_of_float', 'content': '<document version="2.0"><figure><image src="https://static.us.edusercontent.com/files/2Z56brgVDvJpO3IiBxNGmqNl" width="641" height="234.02607823470413"/></figure><paragraph/><paragraph>I am not sure how to grade this. Can anyone help? </paragraph><paragraph>Should I deduct any points?</paragraph></document>', 'document': '\n\nI am not sure how to grade this. Can anyone help? \n\nShould I deduct any points?', 'category': 'A0', 'subcategory': '', 'subsubcategory': '', 'flag_count': 0, 'star_count': 0, 'view_count': 58, 'unique_view_count': 32, 'vote_count': 0, 'reply_count': 2, 'unresolved_count': 1, 'is_locked': False, 'is_pinned': False, 'is_private': False, 'is_endorsed': False, 'is_answered': True, 'is_archived': False, 'is_anonymous': False, 'is_megathread': False, 'anonymous_comments': False, 'approved_status': 'approved', 'created_at': '2021-02-25T20:28:44.243439+11:00', 'updated_at': '2021-03-03T16:27:05.436025+11:00', 'deleted_at': None, 'pinned_at': None, 'anonymous_id': 2491151282, 'vote': 0, 'is_seen': True, 'is_starred': False, 'is_watched': None, 'glanced_at': '2021-03-03T16:27:05.436561+11:00', 'new_reply_count': 0, 'duplicate_title': None, 'answers': [{'id': 659056, 'user_id': 90208, 'course_id': 4757, 'thread_id': 277600, 'parent_id': None, 'editor_id': None, 'type': 'answer', 'kind': 'normal', 'content': '<document version="2.0"><paragraph>Yeah, that\'s some hideous code.  I suspect they didn\'t understand how integer division works and decided to do everything as floats.  Even given that, the formatting is awful.</paragraph><paragraph>Apply this ad hoc deduction under Style: Other:</paragraph><pre>-1: misused float operations instead of int, resulting in expressions that are \n    too hard to read and inconsistently formatted\n</pre><paragraph/></document>', 'document': "Yeah, that's some hideous code.  I suspect they didn't understand how integer division works and decided to do everything as floats.  Even given that, the formatting is awful.\n\nApply this ad hoc deduction under Style: Other:\n\n-1: misused float operations instead of int, resulting in expressions that are \n    too hard to read and inconsistently formatted\n\n\n", 'flag_count': 0, 'vote_count': 0, 'is_endorsed': False, 'is_anonymous': False, 'is_private': False, 'is_resolved': False, 'created_at': '2021-02-26T00:46:43.226813+11:00', 'updated_at': None, 'deleted_at': None, 'anonymous_id': 553740545, 'vote': 0, 'comments': [{'id': 664097, 'user_id': 147342, 'course_id': 4757, 'thread_id': 277600, 'parent_id': 659056, 'editor_id': None, 'type': 'comment', 'kind': 'normal', 'content': '<document version="2.0"><paragraph>Thanks!</paragraph></document>', 'document': 'Thanks!', 'flag_count': 0, 'vote_count': 0, 'is_endorsed': False, 'is_anonymous': False, 'is_private': False, 'is_resolved': False, 'created_at': '2021-02-27T00:32:35.859471+11:00', 'updated_at': None, 'deleted_at': None, 'anonymous_id': 2491151282, 'vote': 0, 'comments': []}]}], 'comments': []}, 'users': [{'id': 147342, 'role': 'user', 'name': 'Boao Dong', 'avatar': None, 'course_role': 'student', 'tutorials': {}}, {'id': 90208, 'role': 'user', 'name': 'Michael Clarkson', 'avatar': None, 'course_role': 'admin', 'tutorials': {}}]}
```

... and clean it up a bit: 
```json
{
   "by":"Boao Dong",
   "title":"use of int_of_float",
   "content":"I am not sure how to grade this. Can anyone help? Should I deduct any points?",
   "created_at":"2021-02-25T20:28:44.243439+11:00",
   "answers":[
      {
         "by":"Michael Clarkson",
         "content":"Yeah, that's some hideous code.  I suspect they didn't understand how integer division works and decided to do everything as floats.  Even given that, the formatting is awful.Apply this ad hoc deduction under Style: Other:-1: misused float operations instead of int, resulting in expressions that are \n    too hard to read and inconsistently formatted\n",
         "created_at":"2021-02-26T00:46:43.226813+11:00",
         "comments":[
            {
               "by":"Boao Dong",
               "content":"Thanks!",
               "created_at":"2021-02-27T00:32:35.859471+11:00",
               "comments":[
                  
               ]
            }
         ]
      }
   ],
   "comments":[
      
   ]
}
```

### CODE REVIEW

File: edstem-api/edstem_api/api.py
Comment (bayoumi17m) [inline edstem-api/edstem_api/api.py:70]: This is something I will probably need to do as a whole but we should go through and document things like we documented [this function](https://github.com/oscarso2000/MyCourseIndex/pull/182/files#diff-55a2e1060a9cc44a865f70144bfde3ae33d01a3fee713f09d4b2cd4432405eb5R21).

File: edstem-api/edstem_api/macros.py
Comment (bayoumi17m) [inline edstem-api/edstem_api/macros.py:1]: Again, we will want to eventually document all of these functions. Did you find most were self explanatory or that you tended to struggle with these having to do lots of tests?

File: edstem-api/edstem_api/macros.py
Comment (bayoumi17m) [inline edstem-api/edstem_api/macros.py:222]: Let's coordinate with @rdongre7 I believe I told her that the spec is:

```json
{   
        "title": "",
	"question": "EdStem / Pazza only field",
	"text": "Non- Piazza/ EdsStem field",
	"student answer": "EdStem / Pazza only field",
	"followups": "EdStem / Pazza only field",
	"docmenttype": "Piazza / Lecture notes / Dscussion",
}
```

However, I am realizing there may be other desired values such as how many instructor endorsed or hearts there are (or whatever). Those are prolly values we also want to maintain.

File: edstem-api/edstem_api/macros.py
Comment (TrueshotBarrage) [reply edstem-api/edstem_api/macros.py:1]: To be completely honest, I forgot that your changes didn't have comments lol. I can start adding some if I understand the function, and we can rebase + recommit to the PR

File: edstem-api/edstem_api/macros.py
Comment (bayoumi17m) [inline edstem-api/edstem_api/macros.py:210]: Minor thing, but we probably don't actually care who said what. We only care about their role. It might be worth changing this in another iteration

**Review by bayoumi17m — CHANGES_REQUESTED** [review]
Overall really solid. I think this is almost perfect. I have some minor things that I think we should coordinate on which is mostly specification material. The docs on functions I wrote I can do, and a doctoring for the collapse / recursion would be helpful. It can be super short.

Additionally, we shouldn't push to master. Rather we should be using the `serverless-arch-dev` type branches for our work until we are ready to release.

File: edstem-api/edstem_api/macros.py
Comment (bayoumi17m) [reply edstem-api/edstem_api/macros.py:1]: It was more a question so that when I edit docstrings later, I know how in depth I should actually go into the process of explanation. I'm thinking most of these are decently obvious but the expected types are ambiguous right now.

**Comment by TrueshotBarrage** [comment]
> Overall really solid. I think this is almost perfect. I have some minor things that I think we should coordinate on which is mostly specification material. The docs on functions I wrote I can do, and a doctoring for the collapse / recursion would be helpful. It can be super short.
> 
> Additionally, we shouldn't push to master. Rather we should be using the `serverless-arch-dev` type branches for our work until we are ready to release.

Sounds good, what do you mean by doctoring for the collapse / recursion?

**Comment by bayoumi17m** [comment]
Sorry, I meant to write docstring*. It autocorrected but just a high level scope of what is being done.

File: edstem-api/edstem_api/macros.py
Comment (TrueshotBarrage) [reply edstem-api/edstem_api/macros.py:1]: Yeah, maybe a one liner for the description

**Comment by TrueshotBarrage** [comment]
I added docstrings to the helper functions as well. But since we should merge it into `serverless-arch-dev` branch, I'll go ahead and rebase it once it's approved and make a new PR.

**Review by oscarso2000 — APPROVED** [review]

**Review by bayoumi17m — APPROVED** [review]
@oscarso2000 DO NOT MERGE. @TrueshotBarrage will move the PR to the new branch and we can merge there.

**Pull request merged**
